### PR TITLE
Bind environment of observeChanges callbacks

### PR DESCRIPTION
**Pull request by zimme**
This will use `bindEnvironment` on `observe` and `observeChanges` callbacks.
This will allow `Meteor.userId()` to be used within these callbacks.

See this conversation for background.
https://github.com/meteor/meteor/pull/8629#issuecomment-300084704

/cc @mitar

### CODE REVIEW

File: packages/mongo/mongo_driver.js [hunk 1]
```diff
@@ -949,7 +949,23 @@ Cursor.prototype.observe = function (callbacks) {
...
+    'changed',
+    'removedAt',
+    'removed',
+    'movedTo'
+  ];
   var ordered = LocalCollection._observeChangesCallbacksAreOrdered(callbacks);
+
+  methods.forEach(function (method) {
+    if (callbacks[method] && typeof callbacks[method] == "function") {
+      callbacks[method] = Meteor.bindEnvironment(callbacks[method]);
```
Comment (mitar) [inline packages/mongo/mongo_driver.js]: This could give us opportunity to also handle errors from those callbacks better. Do we want to do anything about that? Stop the publication? Call a callback on the publication? :-)

Comment (mitar) [inline packages/mongo/mongo_driver.js]: (But it is true that not all observes are called from the publication.

**Comment by mitar** [comment]
Are you sure `observe` is handled? That was just my memory from the past. But if you checked, then great.

**Comment by mitar** [comment]
Does currently `observeChanges` have to be called from a Fiber?

File: packages/mongo/mongo_driver.js (hunk 1 above)
Comment (zimme) [inline packages/mongo/mongo_driver.js]: I don't think we should tie this to publications in that we automatically stop publications. People can just use `self.stop()` or arrow functions and `this.stop()` to do that on their own?

**Comment by zimme** [comment]
> Are you sure observe is handled? That was just my memory from the past. But if you checked, then great.

Yep, `observe` uses `observeChanges` internally after a detour through `LocalCollection` first.

> Does currently observeChanges have to be called from a Fiber?

Not sure, will look into this.

File: packages/mongo/mongo_driver.js (hunk 1 above)
Comment (mitar) [inline packages/mongo/mongo_driver.js]: No, I was more thinking of providing a new callback on publish, like `onError`, and then it gets called. If you want to stop publish on error, you can call `this.stop()` yourself.

Comment (mitar) [inline packages/mongo/mongo_driver.js]: The point is, if observe callbacks throws an error, currently there is no easy way to handle that.

Comment (zimme) [inline packages/mongo/mongo_driver.js]: I started to look into this a bit more but the code gets complex quite quick and I don't feel I have time to dig into that right now so I'd say let's keep this PR down to just binding the environment for the callbacks in the best way possible and open an issue for discussing that a better error handling strategy would be.

**Comment by zimme** [comment]
> Does currently observeChanges have to be called from a Fiber?

Dug into the code a bit more and once i got into this function `addHandleAndSendInitialAdds` it looks like it's using some Future function.

File: packages/mongo/mongo_driver.js [hunk 2]
```diff
@@ -949,7 +949,23 @@ Cursor.prototype.observe = function (callbacks) {
...
   var self = this;
+  var methods = [
+    'addedAt',
+    'added',
+    'changedAt',
+    'changed',
+    'removedAt',
+    'removed',
+    'movedTo'
+  ];
```
Comment (zimme) [inline packages/mongo/mongo_driver.js]: Can we iterate over the properties of callbacks to bindEnvironment without explicitly using a list of methods that needs updating if they ever change? Is this available in some common internal variable maybe?

**Comment by zimme** [comment]
I found this comment while going thought some old issues.
https://github.com/meteor/meteor/issues/853#issuecomment-25515291
It kinds confused me as this is what we're trying to do with this PR and from that comment that was at some point the default?

File: packages/mongo/mongo_driver.js (hunk 2 above)
Comment (mitar) [inline packages/mongo/mongo_driver.js]: In minimongo it is made even uglier: https://github.com/meteor/meteor/blob/d6c4281eebe8fffcdeb415ccc361784f854126f4/packages/minimongo/minimongo.js#L381

I think a list is OK. I would not go wrapping everything because there might be other things there.

**Comment by mitar** [comment]
Hm, crazy question: did you try to access DDP variables inside the observe callbacks? Maybe they are already available? (You have to check both first and non-first run of a callback to be sure.)

I think in that comment the keyword is **implicit**. But I would say that if it does not work, then explicit is better. :-) Just to make sure.

**Comment by zimme** [comment]
> Hm, crazy question: did you try to access DDP variables inside the observe callbacks? Maybe they are already available? (You have to check both first and non-first run of a callback to be sure.)

I've tried to access `DDP._CurrentPublicationInvocation` inside of both `observe` and `observeChanges` callbacks without this patch both on first run and consecutive runs and it's not available.
With this patch, it works as expected. So from what i can tell the environment isn't bound implicitly either. Maybe it was the default at sometime or he was talking about something else in that comment.

**Comment by zimme** [comment]
This PR should be done.

File: History.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 ## v.NEXT
 
+* `observe`/`observeChanges` callbacks are now bound.
```
Comment (mitar) [inline History.md]: Maybe explain that this is on the server side. And what exactly means "bound"? To `this`? Or something else. ;-) (I know what it is, I am just saying that it is unclear for readers.)

BTW, what is `this` bound to inside callbacks? Should we bound them to something?

File: History.md [hunk 4]
```diff
@@ -1,5 +1,10 @@
...
+  The same `EnvironmentVariable`s that was present when
+  `observe`/`observeChanges` was called is now available inside the callbacks.
```
Comment (mitar) [inline History.md]: I think we should explain what binding environment is in documentation (see https://github.com/meteor/docs/issues/139) and then here just say:

>  `observe`/`observeChanges` callbacks have now their Meteor environment bound.

With "Meteor environment bound" linking to description of `Meteor.bindEnvironment`.

File: packages/mongo/mongo_driver.js (hunk 1 above)
Comment (mitar) [inline packages/mongo/mongo_driver.js]: If we are not supporting intercepting errors, maybe then we should at least name the error handler? Like "cursor observe"?

File: History.md (hunk 3 above)
Comment (zimme) [inline History.md]: Will try and elaborate a bit.

`this` isn't set to anything specific, and if we were to change it I don't know to what.

Comment (mitar) [inline History.md]: Crazy idea: to `DDP._CurrentPublicationInvocation.get()`.

File: History.md (hunk 4 above)
Comment (zimme) [inline History.md]: This makes sense, maybe we should hold of on this PR until we have that documentation instead?

File: History.md (hunk 3 above)
Comment (mitar) [inline History.md]: (This would go into effect only if you have not bound your function yourself to some other `this`.)

Comment (zimme) [inline History.md]: What if you're not using `observe`/`observeChanges` from within a `publication`?

Comment (mitar) [inline History.md]: Then it will be `null`.

Comment (zimme) [inline History.md]: I'd say, let's do this in a separate PR. `observe` and `observeChanges` do have separate `this` so jus adding a context to `bindEnvironment` isn't going to set the context for both `observe` and `observeChanges`.

Comment (mitar) [inline History.md]: OK.

File: packages/mongo/mongo_driver.js (hunk 1 above)
Comment (zimme) [inline packages/mongo/mongo_driver.js]: There, used named exception handler. Should be a bit easier to follow the stack trace.

File: packages/mongo/mongo_driver.js (hunk 2 above)
Comment (zimme) [inline packages/mongo/mongo_driver.js]: :+1:

File: History.md [hunk 5]
```diff
@@ -1,5 +1,10 @@
 ## v.NEXT
 
+* `observe`/`observeChanges` callbacks are now bound using `Meteor.bindEnvironment`.
+  The same `EnvironmentVariable`s that was present when `observe`/`observeChanges`
+  was called is now available inside the callbacks.
```
Comment (mitar) [inline History.md]: "are now available".

File: History.md [hunk 6]
```diff
@@ -1,5 +1,10 @@
 ## v.NEXT
 
+* `observe`/`observeChanges` callbacks are now bound using `Meteor.bindEnvironment`.
+  The same `EnvironmentVariable`s that was present when `observe`/`observeChanges`
```
Comment (mitar) [inline History.md]: was -> were

Comment (zimme) [inline History.md]: Fixed

**Comment by mitar** [comment]
So who merges? Can I?

**Comment by hwillson** [comment]
We were just waiting for the tests to pass (they were re-running). They failed again, but the failure isn't related to this PR. I've restarted them, so let's give them another change before merging. Thanks!

**Comment by zimme** [comment]
The tests needs to be re-run again, do we know why circle-ci get segmentation faults so often?
